File: lib/zexray/nif/cursor.ex
defmodule Zexray.NIF.Cursor do
  @moduledoc false

  defmacro __using__(_opts) do
    quote do
      @nifs_cursor [
        # Cursor
        show_cursor: 0,
        hide_cursor: 0,
        is_cursor_hidden: 0,
        enable_cursor: 0,
        disable_cursor: 0,
        is_cursor_on_screen: 0
      ]

      ############
      #  Cursor  #
      ############

      @doc """
      Shows cursor

      ```c
      // raylib.h
      RLAPI void ShowCursor(void);
      ```
      """
      @doc group: :cursor
      @spec show_cursor() :: :ok
      def show_cursor(), do: :erlang.nif_error(:undef)

      @doc """
      Hides cursor

      ```c
      // raylib.h
      RLAPI void HideCursor(void);
      ```
      """
      @doc group: :cursor
      @spec hide_cursor() :: :ok
      def hide_cursor(), do: :erlang.nif_error(:undef)

      @doc """
      Check if cursor is not visible

      ```c
      // raylib.h
      RLAPI bool IsCursorHidden(void);
      ```
      """
      @doc group: :cursor
      @spec is_cursor_hidden() :: boolean
      def is_cursor_hidden(), do: :erlang.nif_error(:undef)

      @doc """
      Enables cursor (unlock cursor)

      ```c
      // raylib.h
      RLAPI void EnableCursor(void);
      ```
      """
      @doc group: :cursor
      @spec enable_cursor() :: :ok
      def enable_cursor(), do: :erlang.nif_error(:undef)

      @doc """
      Disables cursor (lock cursor)

      ```c
      // raylib.h
      RLAPI void DisableCursor(void);
      ```
      """
      @doc group: :cursor
      @spec disable_cursor() :: :ok
      def disable_cursor(), do: :erlang.nif_error(:undef)

      @doc """
      Check if cursor is on the screen

      ```c
      // raylib.h
      RLAPI bool IsCursorOnScreen(void);
      ```
      """
      @doc group: :cursor
      @spec is_cursor_on_screen() :: boolean
      def is_cursor_on_screen(), do: :erlang.nif_error(:undef)
    end
  end
end

File: src/nifs/cursor.zig
const std = @import("std");
const assert = std.debug.assert;
const e = @import("../erl_nif.zig");
const rl = @import("../raylib.zig");

const core = @import("../core.zig");

pub const exported_nifs = [_]e.ErlNifFunc{
    // Cursor
    .{ .name = "show_cursor", .arity = 0, .fptr = nif_show_cursor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "hide_cursor", .arity = 0, .fptr = nif_hide_cursor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "is_cursor_hidden", .arity = 0, .fptr = nif_is_cursor_hidden, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "enable_cursor", .arity = 0, .fptr = nif_enable_cursor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "disable_cursor", .arity = 0, .fptr = nif_disable_cursor, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "is_cursor_on_screen", .arity = 0, .fptr = nif_is_cursor_on_screen, .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
};

//////////////
//  Cursor  //
//////////////

/// Shows cursor
///
/// raylib.h
/// RLAPI void ShowCursor(void);
fn nif_show_cursor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.ShowCursor();

    // Return

    return core.Atom.make(env, "ok");
}

/// Hides cursor
///
/// raylib.h
/// RLAPI void HideCursor(void);
fn nif_hide_cursor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.HideCursor();

    // Return

    return core.Atom.make(env, "ok");
}

/// Check if cursor is not visible
///
/// raylib.h
/// RLAPI bool IsCursorHidden(void);
fn nif_is_cursor_hidden(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const is_cursor_hidden = rl.IsCursorHidden();

    // Return

    return core.Boolean.make(env, is_cursor_hidden);
}

/// Enables cursor (unlock cursor)
///
/// raylib.h
/// RLAPI void EnableCursor(void);
fn nif_enable_cursor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EnableCursor();

    // Return

    return core.Atom.make(env, "ok");
}

/// Disables cursor (lock cursor)
/// raylib.h
///
/// RLAPI void DisableCursor(void);
fn nif_disable_cursor(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.DisableCursor();

    // Return

    return core.Atom.make(env, "ok");
}

/// Check if cursor is on the screen
///
/// raylib.h
/// RLAPI bool IsCursorOnScreen(void);
fn nif_is_cursor_on_screen(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) callconv(.C) e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const is_cursor_on_screen = rl.IsCursorOnScreen();

    // Return

    return core.Boolean.make(env, is_cursor_on_screen);
}
